### ISSUE TITLE
% in url params causes website loading to fail.

### ISSUE THREAD

**Issue by Abhirup-99** [gatsbyjs/gatsby#34797]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

Basically title, % in url params causes a gatsby site to fail to load properly.
Culprit 
```
const trimPathname = rawPathname => {
  const pathname = decodeURIComponent(rawPathname)
  // Remove the pathPrefix from the pathname.
  const trimmedPathname = stripPrefix(
    pathname,
    decodeURIComponent(__BASE_PATH__)
  )
    // Remove any hashfragment
    .split(`#`)[0]

  return trimmedPathname
}
```

Error in console.
```
find-path.js:10 Uncaught (in promise) URIError: URI malformed
    at decodeURIComponent (<anonymous>)
    at c (find-path.js:10:20)
    at g (find-path.js:141:27)
    at f (find-path.js:57:27)
    at e (find-path.js:122:19)
    at t.loadPage (loader.js:226:22)
    at Object.loadPage (loader.js:638:33)
    at production-app.js:174:3
 ```

### Reproduction Link

https://www.gatsbyjs.com/?ref=%25
 Our website: https://even.in/?param=%25, fails to load with the above error thrown.

### Steps to Reproduce

Open gatsby website with the above URL.

### Expected Result

It should load fine.

### Actual Result

It throws console errors. The page is non responsive with none of the JS bundles loading.

### Environment

```shell
Any env.
```

**Comment by marvinjude** [comment]
Hi @Abhirup-99,

Thanks for reporting this issue. 

This seems to be throwing `URIError` because `decodeURIComponent` is trying to decode an unexpected character `%`  which is the result of the previous decoding - `decodeURIComponent("%25")`. I'm currently [exploring a fix](https://github.com/gatsbyjs/gatsby/pull/34816). For now, you can try any feasible workaround.

**Comment by Abhirup-99** [comment]
@marvinjude thanks for exploring the fix. We would be moving to Gatsby v4 shortly for SSR and would need this bit unblocked before that. Any ETA would be appreciated.

**Comment by garyforsterio** [comment]
Thanks for opening this issue @Abhirup-99 and also for the speedy investigation @marvinjude. We're also encountering this issue on our production environment and are currently investigating a workaround to temporarily replace % with something else. 

Any ETA for a fix would be really appreciated 🙏

**Comment by marvinjude** [comment]
Hi @garyforsterio, @Abhirup-99  the potential fix I tried out seem to be breaking a test and may not be a good fix. I will have a look at this again to see if there's an easy fix as we're currently putting more effort into other things at this point in time. Will post an update on this later today.

**Comment by Abhirup-99** [comment]
Hi, thanks for the response. I would like to mention this is a regression with gatsby v4. This wasn't the case with Gatsby v3. We are in the process of moving to v4 to take advantage of the SSR functionalities but this is holding us back. Any chance this can get a higher priority.

**Comment by marvinjude** [comment]
Hi @Abhirup-99,  We're sorry that you're blocked on moving to v4. We've bumped up the priority on this and will be addressed in the current sprint.
